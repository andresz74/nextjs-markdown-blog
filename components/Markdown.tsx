import React from 'react';
import ReactMarkdown from 'react-markdown';
import remarkGfm from 'remark-gfm';
import CodeBlock from '@/utils/markdownRenderers/CodeBlock';
import Heading from '@/utils/markdownRenderers/Heading';
import ImageNode from '@/utils/markdownRenderers/ImageNode'
import LinkNode from '@/utils/markdownRenderers/LinkNode';
import ListNode from '@/utils/markdownRenderers/ListNode';
import ListItemNode from '@/utils/markdownRenderers/ListItemNode';
import Paragraph from '@/utils/markdownRenderers/Paragraph';
import Preformatted from '@/utils/markdownRenderers/Preformatted';
import { Components } from 'react-markdown';

interface MarkdownProps {
	children: string;
}

const renderers: Components = {
	a: ({ children, ...props }) => <LinkNode {...props}>{children}</LinkNode>,
	code: ({ className, children, ...props }) => {
		// Block-level code: if `className` is present, it's a fenced code block
		if (className) {
			return <CodeBlock className={className}>{children}</CodeBlock>;
		}
		// Inline code: render without breaking the paragraph
		return <code style={{
			backgroundColor: '#f0f0f0',
			padding: '2px 4px',
			borderRadius: '4px',
			fontFamily: 'monospace',
			fontSize: '0.875rem',
		}} {...props}>{children}</code>;
	},
	h1: ({ children, ...props }) => <Heading level={1} {...props}>{children}</Heading>,
	h2: ({ children, ...props }) => <Heading level={2} {...props}>{children}</Heading>,
	h3: ({ children, ...props }) => <Heading level={3} {...props}>{children}</Heading>,
	h4: ({ children, ...props }) => <Heading level={4} {...props}>{children}</Heading>,
	h5: ({ children, ...props }) => <Heading level={5} {...props}>{children}</Heading>,
	h6: ({ children, ...props }) => <Heading level={6} {...props}>{children}</Heading>,
	img: ({ alt, src, ...props }) => <ImageNode alt={alt ?? ''} src={src ?? ''} {...props} />,
	li: ({ children, ...props }) => <ListItemNode {...props}>{children}</ListItemNode>,
	ol: ({ children, ...props }) => <ListNode ordered={true} {...props}>{children}</ListNode>,
	ul: ({ children, ...props }) => <ListNode ordered={false} {...props}>{children}</ListNode>,
	p: ({ children, ...props }) => <Paragraph {...props}>{children}</Paragraph>,
	pre: ({ children, ...props }) => <Preformatted {...props}>{children}</Preformatted>,
};

const Markdown: React.FC<MarkdownProps> = ({ children }) => {
	return (
		<ReactMarkdown components={renderers} remarkPlugins={[[remarkGfm, { singleTilde: false }]]}>
			{children}
		</ReactMarkdown>
	);
};

export default Markdown;
