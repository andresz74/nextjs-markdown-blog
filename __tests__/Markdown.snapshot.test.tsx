import React from 'react';
import { render } from '@testing-library/react';
import Markdown from '@/components/Markdown';

describe('Markdown snapshot', () => {
  it('renders markdown with headings and text', () => {
    const md = `# Hello\n\nThis is a test of **Markdown**.`;
    const { container } = render(<Markdown>{md}</Markdown>);
    expect(container).toMatchSnapshot();
  });
});
