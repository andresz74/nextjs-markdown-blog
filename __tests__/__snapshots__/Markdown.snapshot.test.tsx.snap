// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown snapshot renders markdown with headings and text 1`] = `
<div>
  <h1>
    Hello
  </h1>
  

  <p>
    This is a test of 
    <strong>
      Markdown
    </strong>
    .
  </p>
</div>
`;
